File: 383 - Ransom Note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
    let map = new Map();
    for (let c of magazine) {
        if (!map.has(c)) {
            map.set(c, 1);
        } else {
            map.set(c, map.get(c) + 1);
        }
    }
    for (let c of ransomNote) {
        if (map.has(c) && map.get(c) > 0) {
            map.set(c, map.get(c) - 1);
        } else {
            return false;
        }
    }
    return true;
};

File: Java Solutions/383 - Ransom Note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {

        int[] char_counts = new int[26];
        for (char c : magazine.toCharArray()) {
            char_counts[c - 'a']++;
        }

        for (char c : ransomNote.toCharArray()) {
            if (--char_counts[c - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }
}